extensions/amp-form/0.1/amp-form.js

```diff
@@ -245,17 +266,113 @@ function onInvalidInputBlur_(event) {
  * @param {!HTMLInputElement} input
  */
 function reportInputValidity(input) {
-  input./*OK*/focus();
+  if (input.reportValidity) {
+    input.reportValidity();
+  } else {
+    input./*OK*/focus();
+
+    // Remove any previous listeners on the same input. This avoids adding many
+    // listeners on the same element when the user submit pressing Enter or any
+    // other method to submit the form without the element losing focus.
+    input.removeEventListener('blur', onInvalidInputBlur_);
+    input.removeEventListener('keyup', onInvalidInputKeyUp_);
+
+    input.addEventListener('keyup', onInvalidInputKeyUp_);
+    input.addEventListener('blur', onInvalidInputBlur_);
+    validationBubble.show(input, input.validationMessage);
+  }
+}
 
-  // Remove any previous listeners on the same input. This avoids adding many
-  // listeners on the same element when the user submit pressing Enter or any
-  // other method to submit the form without the element losing focus.
-  input.removeEventListener('blur', onInvalidInputBlur_);
-  input.removeEventListener('keyup', onInvalidInputKeyUp_);
 
-  input.addEventListener('keyup', onInvalidInputKeyUp_);
-  input.addEventListener('blur', onInvalidInputBlur_);
-  validationBubble.show(input, input.validationMessage);
+/**
+ * Checks user validity for all inputs, fieldsets and the form.
+ * @param {!HTMLFormElement} form
+ */
+function checkUserValidityOnSubmission_(form) {
+  const inputs = form.querySelectorAll('input,select,textarea,fieldset');
+  for (let i = 0; i < inputs.length; i++) {
+    checkUserValidity(inputs[i]);
+  }
+  checkUserValidity(form);
+}
+
+
+/**
+ * Checks user validity which applies .user-valid and .user-invalid AFTER the user
+ * interacts with the input by moving away from the input (blur) or by changing its
+ * value (input).
+ *
+ * See :user-invalid spec for more details:
+ *   https://drafts.csswg.org/selectors-4/#user-pseudos
+ *
+ * The specs are still not fully specified. The current solution tries to follow a common
+ * sense approach for when to apply these classes. As the specs gets clearer, we should
+ * strive to match it as much as possible.
+ *
+ * @param {!HTMLInputElement|!HTMLFormElement|!HTMLFieldSetElement} element
+ * @returns {boolean} true to indicate that validity should propagate to ancestors.
+ */
+function checkUserValidity(element) {
+  const currentUserValid = element.classList.contains('user-valid');
+  const currentUserInvalid = element.classList.contains('user-invalid');
```

Comment:
Using an enum instead of double booleans might make this a little easier to follow. As I understand it, you can either be `currentUserValid`, `currentUserInvalid`, or neither. But, double booleans allow for a 4th condition where both are true, making it a bit more difficult to reason about.